controller/organization.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/goadesign/goa"
+	"github.com/satori/go.uuid"
+	"strings"
+)
+
+const ORGANIZATION_OWNER_ROLE = "owner"
+
+// OrganizationController implements the organization resource.
+type OrganizationController struct {
+	*goa.Controller
+	db           application.DB
+	TokenManager token.Manager
+}
+
+// NewOrganizationController creates a organization controller.
+func NewOrganizationController(service *goa.Service, db application.DB) *OrganizationController {
+	return &OrganizationController{Controller: service.NewController("OrganizationController"), db: db}
+}
+
+// Create runs the create action.
+func (c *OrganizationController) Create(ctx *app.CreateOrganizationContext) error {
+	currentUser, err := login.ContextIdentity(ctx)
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrUnauthorized(err.Error()))
+	}
+
+	if len(strings.TrimSpace(*ctx.Payload.Name)) == 0 {
+		return jsonapi.JSONErrorResponse(ctx, goa.ErrBadRequest("Organization name cannot be empty"))
+	}
+
+	var organizationId uuid.UUID
+
+	err = application.Transactional(c.db, func(appl application.Application) error {
+
+		// Lookup the identity for the current user
+		userIdentity, err := appl.Identities().Load(ctx, *currentUser)
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "Error looking up current user")
+		}
+
+		// Lookup the organization resource type
+		resourceType, err := appl.ResourceTypeRepository().Lookup(ctx, account.IDENTITY_RESOURCE_TYPE_ORGANIZATION)
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "Error looking up resource type 'identity/organization'")
+		}
+
+		// Create the organization resource
+		res := &resource.Resource{
+			Name:           *ctx.Payload.Name,
+			ResourceType:   *resourceType,
+			ResourceTypeID: resourceType.ResourceTypeID,
+		}
+
+		err = appl.ResourceRepository().Create(ctx, res)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
+		}
+
+		// Create the organization identity
+		orgIdentity := &account.Identity{
+			IdentityResourceID: &res.ResourceID,
+		}
+
+		err = appl.Identities().Create(ctx, orgIdentity)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
+		}
+
+		organizationId = orgIdentity.ID
+
+		// Lookup the identity/organization owner role
+		ownerRole, err := appl.RoleRepository().Lookup(ctx, ORGANIZATION_OWNER_ROLE, account.IDENTITY_RESOURCE_TYPE_ORGANIZATION)
+
+		if err != nil {
+			return errors.NewInternalErrorFromString(ctx, "Error looking up owner role for 'identity/organization' resource type")
+		}
+
+		// Assign the owner role for the new organization to the current user
+		identityRole := &role.IdentityRole{
+			IdentityID: userIdentity.ID,
+			ResourceID: res.ResourceID,
+			RoleID:     ownerRole.RoleID,
+		}
+
+		err = appl.IdentityRoleRepository().Create(ctx, identityRole)
+		if err != nil {
+			return jsonapi.JSONErrorResponse(ctx, goa.ErrInternal(err.Error()))
```

Comment:
We typically use `errors.NewInternalError(..)` or `errors.NewInternalErrorFromString(..)` from our `errors` packages.

Reply:
I've gone through and updated these - there is one instance where I use the following:

		return jsonapi.JSONErrorResponse(ctx, goa.ErrBadRequest("Organization name cannot be empty"))

The errors.NewBadParameterError() method doesn't seem to be capable of accepting a msg parameter, so is there another way of returning this error with a message or is the above code the only way to do this?

Reply:
`errors.NewBadParameterError("organization","empty")` not the cleanest way, but this could be done. I don't have a very strong opinion on this - so you could do either ways :)